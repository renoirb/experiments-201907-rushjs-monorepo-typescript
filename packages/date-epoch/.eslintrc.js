/* eslint-disable @typescript-eslint/no-var-requires */
const base = require('@renoirb/conventions-use-eslint')

const main = {
  ...base,
  parserOptions: {
    tsconfigRootDir: __dirname,
    project: './tsconfig.eslint.json',
  },
  rules: {
    ...base.rules,
    // See comments in @renoirb/tools-bundling-helpers .eslintrc.js
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/no-explicit-any': 'off',
    '@typescript-eslint/typedef': 'off',
  },
}

module.exports = main
