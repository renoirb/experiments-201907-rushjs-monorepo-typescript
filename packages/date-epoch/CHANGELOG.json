{
  "name": "@renoirb/date-epoch",
  "entries": [
    {
      "version": "1.1.0",
      "tag": "@renoirb/date-epoch_v1.1.0",
      "date": "Mon, 29 Jun 2020 02:44:25 GMT",
      "comments": {
        "minor": [
          {
            "comment": "Upgrading deps, minor refactor"
          },
          {
            "comment": "Make DateEpoch behave more closely to Date"
          }
        ],
        "patch": [
          {
            "comment": "Reworked DateEpoch signature and support coercion from string, number, number in a string, and new annotation"
          }
        ],
        "dependency": [
          {
            "comment": "Updating dependency \"@renoirb/conventions-use-bili\" from `1.1.0-dev.7` to `1.1.0`"
          },
          {
            "comment": "Updating dependency \"@renoirb/conventions-use-eslint\" from `1.1.0-dev.7` to `1.1.0`"
          },
          {
            "comment": "Updating dependency \"@renoirb/conventions-use-jest\" from `1.1.0-dev.7` to `1.1.0`"
          },
          {
            "comment": "Updating dependency \"@renoirb/conventions-use-prettier\" from `1.1.0-dev.7` to `1.1.0`"
          },
          {
            "comment": "Updating dependency \"@renoirb/conventions-use-typescript-3\" from `1.1.0-dev.7` to `1.1.0`"
          },
          {
            "comment": "Updating dependency \"@renoirb/tools-bundling-helpers\" from `1.1.0-dev.7` to `1.1.0`"
          }
        ]
      }
    },
    {
      "version": "1.0.2",
      "tag": "@renoirb/date-epoch_v1.0.2",
      "date": "Thu, 02 Jan 2020 23:25:02 GMT",
      "comments": {
        "patch": [
          {
            "comment": "Make TypeScript and TSLib peerDependency"
          }
        ],
        "dependency": [
          {
            "comment": "Updating dependency \"@renoirb/conventions-use-bili\" from `^1.0.1` to `^1.0.2`"
          },
          {
            "comment": "Updating dependency \"@renoirb/conventions-use-typescript-3\" from `^1.0.1` to `^1.0.2`"
          },
          {
            "comment": "Updating dependency \"@renoirb/tools-bundling-helpers\" from `^1.0.1` to `^1.0.2`"
          }
        ]
      }
    }
  ]
}
