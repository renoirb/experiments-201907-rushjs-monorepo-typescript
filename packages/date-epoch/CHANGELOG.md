# Change Log - @renoirb/date-epoch

This log was last generated on Mon, 17 Aug 2020 02:57:48 GMT and should not be manually modified.

## 1.2.0
Mon, 17 Aug 2020 02:57:48 GMT

### Minor changes

- Browser package is now IIFE

## 1.1.3
Mon, 03 Aug 2020 23:53:08 GMT

*Version update only*

## 1.1.2
Tue, 14 Jul 2020 03:23:31 GMT

### Patches

- Upgrading deps

## 1.1.1
Mon, 29 Jun 2020 04:36:05 GMT

### Patches

- ESM/MJS module

## 1.1.0
Mon, 29 Jun 2020 02:44:25 GMT

### Minor changes

- Upgrading deps, minor refactor
- Make DateEpoch behave more closely to Date

### Patches

- Reworked DateEpoch signature and support coercion from string, number, number in a string, and new annotation

## 1.0.2
Thu, 02 Jan 2020 23:25:02 GMT

### Patches

- Make TypeScript and TSLib peerDependency

