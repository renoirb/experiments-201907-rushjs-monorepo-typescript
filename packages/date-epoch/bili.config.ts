import useBili from '@renoirb/conventions-use-bili'

const main = useBili(
  {
    input: 'src/index.ts',
    bundleNodeModules: true,
    output: {
      format: ['cjs', 'iife', 'esm'],
    },
  },
  { firstYear: 2003 },
)

export default main
