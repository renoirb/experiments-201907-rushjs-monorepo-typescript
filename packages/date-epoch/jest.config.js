/* eslint-disable @typescript-eslint/no-var-requires */
const base = require('@renoirb/conventions-use-jest')

const main = {
  ...base,
  setupFiles: ['jest-date-mock'],
  prettierPath: '<rootDir>/node_modules/@renoirb/conventions-code-formatter/node_modules/prettier/bin-prettier.js',
}

module.exports = main
