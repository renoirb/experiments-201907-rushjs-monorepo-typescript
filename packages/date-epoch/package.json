{
  "name": "@renoirb/date-epoch",
  "version": "1.1.3",
  "description": "Date internal representation utilities using Numbers of seconds since January 1st 1970 (a.k.a. Epoch)",
  "homepage": "https://github.com/renoirb/experiments-201908-rush-typescript-just-bili-monorepo/tree/v1.x-dev/packages/date-epoch",
  "repository": {
    "type": "git",
    "url": "https://github.com/renoirb/experiments-201908-rush-typescript-just-bili-monorepo.git",
    "directory": "packages/date-epoch"
  },
  "license": "MIT",
  "author": {
    "name": "Renoir Boulanger",
    "email": "contribs@renoirboulanger.com"
  },
  "exports": {
    "browser": "dist/index-browser.js",
    "import": "dist/index.mjs",
    "require": "dist/index.js"
  },
  "main": "dist/index.js",
  "jsdelivr": "dist/index-browser.js",
  "module": "dist/index.mjs",
  "typings": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "use-cross-env use-run-all clean build:*",
    "build:browser": "use-cross-env use-bili -c bili.config.ts --target browser --format umd --module-name DateEpoch --input src/index-browser.ts --file-name index-browser.js",
    "build:node": "use-cross-env use-bili -c bili.config.ts --target node --format cjs --file-name index.js",
    "build:node-mjs": "use-cross-env use-bili -c bili.config.ts --target node --format esm --file-name index.mjs",
    "clean": "use-cross-env use-rimraf .rpt2_cache dist *.*.log",
    "fix": "use-cross-env use-prettier '**/*.{ts,md,json,js}' --write",
    "lint": "use-cross-env use-eslint --fix 'src/**/*.ts'",
    "prepublishOnly": "rushx build",
    "test": "use-cross-env use-jest --detectOpenHandles",
    "test:snapshots": "use-cross-env use-jest --detectOpenHandles -u"
  },
  "devDependencies": {
    "@babel/core": "^7.10.1",
    "@babel/plugin-transform-runtime": "^7.10.1",
    "@babel/preset-env": "^7.10.2",
    "@babel/preset-typescript": "^7.10.1",
    "@babel/runtime-corejs3": "^7.10.1",
    "@renoirb/conventions-use-bili": "1.1.2",
    "@renoirb/conventions-use-eslint": "workspace:*",
    "@renoirb/conventions-use-jest": "workspace:*",
    "@renoirb/conventions-use-prettier": "workspace:*",
    "@renoirb/conventions-use-typescript-3": "workspace:*",
    "@renoirb/tools-bundling-helpers": "1.1.2",
    "@types/jest": "^25.2.3",
    "@types/node": "^10.17.13",
    "core-js": "^3.6.0",
    "jest-date-mock": "^1.0.8",
    "typescript": "~3.9.5"
  }
}
