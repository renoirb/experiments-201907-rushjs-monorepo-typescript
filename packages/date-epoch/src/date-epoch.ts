import { toDate, timestamp, coerceMilliseconds } from './helpers'

/**
 * DateEpoch: An utility to convert UNIX Epoch Integer into a Date Object.
 *
 * Helpers to work with UNIX Epoch notation and JavaScript Date objects.
 *
 * This should be an almost exact copy of [1], notice how strict typing had been added.
 *
 * [1]: https://bitsrc.io/renoirb/bindings/types/date-epoch/code
 */
export class DateEpoch {
  private readonly _epoch: number

  get epoch(): number {
    return this._epoch
  }

  /**
   * Constructor: Optimistically take a number and try to create a Date object.
   *
   * @param epoch Hopefully a Number or a String with a Number, hopefully an UNIX Epoch.
   */
  public constructor(epoch?: any) {
    let value: number
    switch (true) {
      case Number.isFinite(epoch):
        value = coerceMilliseconds(epoch)
        break

      case epoch && 'getTime' in epoch:
        value = epoch.getTime()
        break

      default:
        value = timestamp()
        break
    }
    this._epoch = value
    Object.defineProperty(this, '_epoch', {
      value,
      writable: false,
      configurable: false,
      enumerable: false,
    })
  }

  public toJSON(): number {
    return this._epoch
  }

  public toString(): string {
    return String(this._epoch)
  }

  /**
   * Convert internal epoch into a native Date object.
   *
   * @returns Date
   */
  public toDate(): Date {
    return toDate(this._epoch)
  }
}
