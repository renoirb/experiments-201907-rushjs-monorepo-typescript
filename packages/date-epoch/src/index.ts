import { DateEpoch } from './date-epoch'

export * from './decorators'

export { default as helpers, DateRangeTuple, FuzzyDateType } from './helpers'

export { DateEpoch }

export default DateEpoch
