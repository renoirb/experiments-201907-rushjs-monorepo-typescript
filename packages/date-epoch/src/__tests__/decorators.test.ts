// import { epoch } from '../decorators'
import { DateEpoch } from '../date-epoch'

const inputEpochInt = 1_533_762_055_251
const expectEpochIntToISOString = '2018-08-08T21:00:55.251Z'
const expectEpochDateAsSTring =
  'Wed Aug 08 2018 17:00:55 GMT-0400 (Eastern Daylight Time)'

class SomeClass {
  // Yeah, decorators would maybe add too much indirection
  // @epoch
  public lastLogin = inputEpochInt
  // if we compare to ECMAScript getters, much clearer of the intent.
  // ... less "magic"
  get lastLoginEpoch(): DateEpoch {
    return new DateEpoch(this.lastLogin)
  }
}

describe('decorators', () => {
  test('integration', () => {
    const subject = new SomeClass()
    expect(subject).toMatchObject({ lastLogin: inputEpochInt })
    expect(subject).toHaveProperty('lastLogin', inputEpochInt)
    expect(subject).toHaveProperty('lastLoginEpoch')
    expect(subject).toMatchSnapshot()
    expect(subject.lastLoginEpoch).toBeInstanceOf(DateEpoch)
    expect(String(subject.lastLoginEpoch.toDate())).toBe(
      expectEpochDateAsSTring,
    )
    expect(JSON.stringify(subject.lastLoginEpoch.toDate())).toMatch(
      expectEpochIntToISOString,
    )
    expect(subject.lastLoginEpoch).toMatchObject(new DateEpoch(inputEpochInt))
  })
})
