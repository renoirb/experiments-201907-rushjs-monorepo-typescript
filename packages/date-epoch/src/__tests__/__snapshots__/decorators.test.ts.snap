// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorators integration 1`] = `
SomeClass {
  "lastLogin": 1533762055251,
}
`;
